sync2/handler2/handler.go

```diff
@@ -433,7 +437,25 @@ func (h *Handler) UpdateUnreadCounts(ctx context.Context, roomID, userID string,
 }
 
 func (h *Handler) OnAccountData(ctx context.Context, userID, roomID string, events []json.RawMessage) {
-	data, err := h.Store.InsertAccountData(userID, roomID, events)
+	// duplicate suppression for multiple devices on the same account.
+	// We suppress by remembering the last bytes for a given account data, and if they match we ignore.
+	dedupedEvents := make([]json.RawMessage, 0, len(events))
+	for i := range events {
+		evType := gjson.GetBytes(events[i], "type").Str
+		key := fmt.Sprintf("%s|%s|%s", userID, roomID, evType)
+		thisHash := fnvHash(events[i])
+		last, _ := h.accountDataMap.Load(key)
+		if last != nil {
+			lastHash := last.(uint64)
+			if lastHash == thisHash {
+				continue // skip this event
+			}
+		}
+		dedupedEvents = append(dedupedEvents, events[i])
+		h.accountDataMap.Store(key, thisHash)
+	}
+
+	data, err := h.Store.InsertAccountData(userID, roomID, dedupedEvents)
```

Comment:
Suppose that E1, E2, ... are all account data events with the **same type**. This logic will behave as follows:

<table>
<th>
 <td>Input
 <td>dedupedEvents
<tr>
 <td>E1
 <td>E1
<tr>
 <td>E1, E1
 <td>E1
<tr>
 <td>E1, E2
 <td>E1, E2
<tr>
 <td>E1, E2, E1
 <td>E1, E2, E1
<tr>
 <td>E1, E1, E2, E2, E1
 <td>E1, E2, E1
</table>

I think the last three are slightly surprising since you only need to insert the most recent event with a given type. (Last write wins, per-type.)

Having written this up, I now see that there is logic for this in AccountDataTable.Insert. So maybe that's fine. (Though it's not clear to me what the responsibilities of the Handler versus the Storage versus the AccountDataTable are.)

Reply:
Yeah we already dedupe when there are dupes in the array in the storage layer, because it broke the SQL when we didn't. This is subtly different as it's across mutliple callbacks i.e it needs some in-memory structure. The handler is the obvious place for this, whereas the storage layer would be surprising.